<div class="alert alert-block alert-info">
Singapore Management University<br>
CS105 Statistical Thinking for Data Science, 2023/24 Term 2
</div>

# CS105 Group Project Submission (Final)

-----
Provide your team details, including section, team number, team members, and the name of the dataset. 
Complete all of the following sections. For any part requiring code to derive your answers, please create a code cell immediately below your response and run the code.
To edit any markdown cell, double click the cell; after editing, execute the markdown cell to collapse it.

Include both Part I and Part II to produce a self-contained notebook. You may fine-tune Part I codes and/or findings based on feedback received on Part I submission.
<br>
-----

## Declaration

<span style="color:red">By submitting this notebook, we declare that **no part of this submission is generated by any AI tool**. We understand that AI-generated submissions will be considered as plagiarism, and just like other plagirisum cases, disciplinary actions will be imposed.</span>

#### Section: G1
#### Team: Group 1
#### Members:
1. CHONG SHIH RUI ZACHARY
2. YANG ZHONG HNG REUBEN
3. KEAGAN PAUL POULOSE
4. LUCAS SONG KANGZHENG

#### Dataset: Wine dataset

## Part I: Exploratory Data Analysis (EDA) [8% of final grade]

In [12]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import OrdinalEncoder

In [13]:
df = pd.read_csv('wine.csv')

In [14]:
df.head()

ID  Fixed_Acidity  Volatile_Acidity  Citric_Acid Residual_Sugar  Chlorides  \
0   1            7.0              0.27         0.36           High      0.045   
1   2            6.3              0.30         0.34            Low      0.049   
2   3            8.1              0.28         0.40            Low      0.050   
3   4            7.2              0.23         0.32       Moderate      0.058   
4   5            7.2              0.23         0.32       Moderate      0.058   

   Free_SO2  Total_SO2    Density      pH  Sulphates Alcohol  Quality  
0      45.0      170.0      >1.00  Acidic       0.45      8%        6  
1      14.0      132.0  0.99-1.00  Acidic       0.49      9%        6  
2      30.0       97.0  0.99-1.00  Acidic       0.44     10%        6  
3      47.0      186.0  0.99-1.00  Acidic       0.40      9%        6  
4      47.0      186.0  0.99-1.00  Acidic       0.40      9%        6

### 1. Overview of dataset [15% of Part I]

**a.** Summarise the background of the dataset [limited to 50 words]

**This dataset gives the quality of wine.
Each row is an observation and represents a type of wine, and along each row, you find values that describe certain attributes about the wine. The Quality variable serves as our target variable that we will be predicting.**

**b.** State the size of the dataset

In [15]:
n_rows, n_columns = df.shape
print(n_rows)
print(n_columns)
print(df.shape)

4898
13
(4898, 13)


**c.** For each variable, describe what it represents and its data type (numerical or categorical)

|Column name|Description|Data type|
|---|---|---|
|ID| Unique Id of a wine | Categorical |
|Fixed_Acidity| Amount of Tartaric acid, measured in g/dm3 | Numerical |
|Volatile_Acidity| Amount of Acetic acid, measured in g/dm3  | Numerical |
|Citric_Acid| Amount measured in g/dm3 | Numerical |
|Residual_Sugar| Level of residual sugar content | Categorical (“High”, “Moderate”, “Low”) |
|Chlorides| Amount of Sodium chloride, measured in g/dm3 | Numerical |
|Free_SO2| Amount measured in mg/dm3 | Numerical |
|Total_SO2| Amount measured in mg/dm3  | Numerical |
|Density| Amount measured in g/dm3 | Categorical (“<0.99”, “0.99-1.00”, “>1.00”) |
|pH| The measure of acidity of the wine | Categorical (“Neutral”, “Acidic”, “Highly Acidic”)| 
|Sulphates|Potassium sulphate, measured in g/dm3 | Numerical |
|Alcohol| Level of alcohol content | Categorical (8% through 14% in discrete levels)|
|Quality| Score based on sensory data | Categorical (0 through 10 (0 is bad, 10 is good))|

### 2. Data pre-processing [35% of Part I]

**a.** For each variable, determine the percentage of missing data. For any column with missing data, describe how you resolve the issue. Clearly state any assumption you made.

In [16]:
n_rows, n_cols = df.shape #Number of rows and columns in wine dataset
(1 - df.count() / n_rows)*100 #Percentage of missing data

ID                  0.000000
Fixed_Acidity       4.450796
Volatile_Acidity    1.020825
Citric_Acid         0.000000
Residual_Sugar      0.000000
Chlorides           0.000000
Free_SO2            0.000000
Total_SO2           0.000000
Density             4.573295
pH                  0.000000
Sulphates           0.000000
Alcohol             0.000000
Quality             0.000000
dtype: float64

|Column name|Percentage of missing data|How to resolve|
|---|---|---|
|ID| 0% | NA |
|Fixed_Acidity| 4.450796% | Less than 5%, so we should impute |
|Volatile_Acidity| 1.020825%  | Less than 5%, so we should impute |
|Citric_Acid| 0% | NA |
|Residual_Sugar| 0% | NA |
|Chlorides| 0% | NA |
|Free_SO2| 0% | NA |
|Total_SO2| 0%  | NA |
|Density| 4.573295% | Less than 5%, so we should impute |
|pH| 0% | NA | 
|Sulphates|0%| NA |
|Alcohol| 0% | NA |
|Quality| 0% | NA |

**Since we are possibly dropping outlier values and the percentage of missing values for each of the respectives columns is <5%, we intend to deal with missing values via imputation. Further, the dataset is small and we do not want to lose too much data by dropping missing values.**

**For Fixed Acidity and Volatile Acidity, mean imputation is preferred as these are numerical variables. For Density, mode imputation is preferred as it is a categorical variable**

In [17]:
# This cell handles the imputation of values for Fixed Acidity, Volatile Acidity and Density columns

fixed_acidity_mean = df.Fixed_Acidity.mean()
volatile_acidity_mean = df.Volatile_Acidity.mean()
density_mode = df.Density.mode()[0]

df.Density.fillna(density_mode, inplace=True)
df.Fixed_Acidity.fillna(fixed_acidity_mean, inplace=True)
df.Volatile_Acidity.fillna(volatile_acidity_mean, inplace=True)
df.count()

ID                  4898
Fixed_Acidity       4898
Volatile_Acidity    4898
Citric_Acid         4898
Residual_Sugar      4898
Chlorides           4898
Free_SO2            4898
Total_SO2           4898
Density             4898
pH                  4898
Sulphates           4898
Alcohol             4898
Quality             4898
dtype: int64

**b.** For each variable, identify outliers (if any) and describe how you resolve the issue. Clearly state any assumption you made.

**Response.** 

**c.** For categorical variables, perform the necessary encoding.

**Response.** 

### 3.	Exploratory analysis and visualization [50% of Part I]

**a.** For each variable, provide relevant summary statistics

In [ ]:
#Five-number summary of numerical measures
df[ ["Fixed_Acidity","Volatile_Acidity","Citric_Acid","Chlorides","Free_SO2","Total_SO2","Sulphates","Quality"] ].describe()

In [ ]:
def category_summary(df, categories):
    for category in categories:
        print(category)
        print()
        # Count of each category
        category_counts = df[category].value_counts()

        # Frequency of each category
        category_frequency = df[category].value_counts(normalize=True)

        # Mode (most frequent category)
        mode_category = df[category].mode()[0]

        # Unique categories
        unique_categories = df[category].nunique()

        # Relative frequency of each category
        relative_frequency = df[category].value_counts(normalize=True)

        print("Count:")
        print(category_counts)
        print("Frequency:")
        print(category_frequency)
        print("Mode (most frequent category):", mode_category)
        print("Unique categories:", unique_categories)
        print("Relative frequency:")
        print(relative_frequency)
        print()

categories = ['Residual_Sugar','Density','pH','Alcohol']

category_summary(df,categories)

We are aware that by adding additional features due to our encoding, we have expanded the size of our dataset. We are also wary of the curse of dimensionality, which is also why we have chosen to use ordinal encoding for some features, and one-hot encoding for others. Nonetheless, we choose to proceed with this form of pre-processing first, and we will tailor the model as necessary once we begin the training stage to see which forms of pre-processing produce the best results.

**b.** For each variable, provide an appropriate visualisation depicting the distribution of its values, and summarize any key observation(s) you made.

**Response.** 

**c.** Perform bi-variate analysis on the variables. You do not need to present the analysis of every pair of variables; only focus on the pairs you believe are worth investigating and explain. For each pair, describe the relationship between the two variables. Use appropriate statistical methods and/or visualizations.

**Response.** 

## Part II: Modeling [12% of final grade]

### 1. Problem formulation [15% of Part II]

**a.** Formulate one regression problem and one classification based on the dataset, in **no more than 50 words**.

**Regression: Wine collectors want to predict the quality of the wine they are purchasing based on its characteristics to achieve a higher return on investment.**

**Classification: Wine collectors want to know which quality band the wine belongs to so that they will make an informed decision when purchasing/auctioning wines.** 

**b.** State which problem (regression or classification) you would be investigating and why, in **no more than 20 words**.

**Response.** 

**c.** Clearly specify the dependent variable you are predicting, and its significance, in **no more than 20 words**

**The dependent variable we are predicting is the quality of the wine, which is directly proportionate to its market value.** 

### 2. Model training [30% of Part II]

**a.** Perform feature selection. For each variable, decide if you want to include it as a feature and provide a justification. You may leverage on your analysis in Part I: EDA and/or perform additional analysis.

**Response.** 

**b.** Split the dataset into train and test sets. Describe how you split step by step.

**Response.** 

**c.** State the model(s) you will train, and explain your choice(s), in **no more than 50 words per model**. You only need to
train one model, but if you do train more models, limit yourself to no more than three---Grading is based on the validity and soundness of your model, rather than the quantity.

**Response.** 

**d.** For each model, perform the training, and report the trained parameters and the training scores, if applicable. 

**Response.** 

### 3. Model evaluation and selection [30% of Part II]

**a.** For each model, predict the response variable on the test set.

**Response.** 

**b.** Describe the metric you use to evaluate your model(s). Report the test scores for each model.

**Response.** 

**c.** If you trained more than one model, identify the final model you would choose for the prediction task, and explain your choice, **in no more than 50 words**.

**Response.** 

### 4. Findings and conclusion [20% of Part II]

**a.** Interpret what your model is implying, and summarize any insight you have drawn from the project. Explain if it is consistent with intuition, and if not, provide a plausible justification. Limit your entire response to **50 words**.

**Response.** 

**b.** Share any lesson you have learned from the project, in **no more than 50 words**.

**Response.** 

### 5. Non-technical protocol [5% of Part II]

**a.** Describe the detailed contribution of each team member, including both the tangible (e.g., implementation, testing, writing) and intangible (e.g., generating ideas, planning, leadership) efforts.

|Team member|Contribution|
|---|---|
|CHONG SHIH RUI ZACHARY| Wrote first draft of Part I (1. Overview of dataset, 2. Data pre-processing), Part II (1. Problem formulation, 5. Non-technical protocol)| 
||Created scatter plot, boxplot and crosstab to conduct bivariate analysis|
||Implemented preprocessing function to normalize numerical columns|
|YANG ZHONG HNG REUBEN| <to fill> | 
||<to fill>|
|KEAGAN PAUL POULOSE| <to fill>  | 
||<to fill>|
|LUCAS SONG KANGZHENG| <to fill> | 
||<to fill>|

**b.** List any references and sources you have cited.

**Response.** 